Mathlib/RingTheory/Polynomial/Hilbert.lean

```diff
@@ -0,0 +1,93 @@
+  terms of `λ` is the formal power series `P(M, λ, X) = Σₙλ(Mₙ)Xⁿ`. The Hilbert-Serre Theorem states
+  that `P(M, λ, X)` can be written as `p(X)/∏ᵢ₌₁,...,ₛ(1 - X ^ dᵢ)`, where `p(X)` is a polynomial
+  with coefficients in `ℤ`.
+* For the case when `d₁,...,dₛ = 1`, the Poincaré series of `M` with respect to `λ` can be expressed
+  as `p(X)/(1 - X)ˢ`. Hence the fact stated in the beginning guarantees an `h : ℚ[X]` such that for
+  any large enough `n : ℕ`, the coefficient of `Xⁿ` in `P(M, λ, X)` equals `h.eval (n : ℚ)`. This
+  `h` is called the Hilbert polynomial of `M` in terms of `λ`, which is what we eventually want to
+  formalise.
+-/
+
+open BigOperators Nat PowerSeries
+
+namespace Polynomial
+
+section greatestFactorOneSubNotDvd
+
+variable {R : Type*} [CommRing R] [Nontrivial R] [NoZeroDivisors R]
+variable (p : R[X]) (hp : p ≠ 0) (d : ℕ)
+
+/--
+Given a polynomial `p`, the factor `f` of `p` such that the product of `f` and
+`(1 - X : R[X]) ^ p.rootMultiplicity 1` equals `p`. We define this here because if `p` is divisible
+by `1 - X`, then the expression `p/(1 - X)ᵈ` can be reduced. We want to construct the Hilbert
+polynomial based on the most reduced form of the fraction `p/(1 - X)ᵈ`. Later we will see that this
+method of construction makes it much easier to calculate the specific degree of the Hilbert
+polynomial.
+-/
+noncomputable def greatestFactorOneSubNotDvd : R[X] :=
+  ((- 1 : R[X]) ^ p.rootMultiplicity 1) *
+  (exists_eq_pow_rootMultiplicity_mul_and_not_dvd p hp 1).choose
+
+end greatestFactorOneSubNotDvd
+
+/--
+A polynomial which makes it easier to define the Hilbert polynomial. See also the theorem
+`Polynomial.preHilbert_eq_choose_sub_add`, which states that for any `d k n : ℕ` with `k ≤ n`,
+`(Polynomial.preHilbert d k).eval (n : ℚ) = (n - k + d).choose d`.
+-/
+noncomputable def preHilbert (d k : ℕ) : ℚ[X] :=
+  (d.factorial : ℚ)⁻¹ • (∏ i : Finset.range d, (X - (C (k : ℚ)) + (C (i : ℚ)) + 1))
+
+theorem preHilbert_eq_choose_sub_add (d k n : ℕ) (hkn : k ≤ n):
+    (preHilbert d k).eval (n : ℚ) = (n - k + d).choose d := by
+  delta preHilbert; simp only [Finset.univ_eq_attach, map_natCast, eval_smul, smul_eq_mul]
+  rw [eval_prod, Finset.prod_attach _ (fun _ => eval _ (_ - _ + _ + _)), add_choose,
+    cast_div (factorial_mul_factorial_dvd_factorial_add ..) (cast_ne_zero.2 <| mul_ne_zero
+    (n - k).factorial_ne_zero d.factorial_ne_zero), cast_mul, div_mul_eq_div_div, div_eq_inv_mul,
+    mul_eq_mul_left_iff, ← cast_div (factorial_dvd_factorial <| Nat.le_add_right (n - k) d)
+    (cast_ne_zero.2 <| factorial_ne_zero (n - k)), ← ascFactorial_eq_div]
+  simp_rw [eval_add, eval_sub, eval_X, eval_natCast, eval_one, ascFactorial_eq_prod_range,
+    cast_prod, cast_add, cast_one, Nat.cast_sub hkn, add_assoc, add_comm, true_or]
+
+/--
+Given `p : ℤ[X]` and `d : ℕ`, the Hilbert polynomial of `p` and `d`. Later we will
+show that `PowerSeries.coeff ℤ n (p * (PowerSeries.invOneSubPow ℤ d))` is equal to
```

Comment:
I think it would be better to get this result in this PR as well, to make sure that this is actually the right definition.

Reply:
I have already proved the statement in the Visual Studio Code of my computer, so I can guarantee that the definition has no problem. But the proof is very long. I'm afraid that if I also include the proof in this PR, then it may not be so easy to be merged to Mathlib 4. I plan to include the proof of the statement in some other PR.

But that is only my own opinion. I'm not sure if my plan is good. What's your opinion?

Reply:
> I think it would be better to get this result in this PR as well, to make sure that this is actually the right definition.

I have already proved the statement in the Visual Studio Code of my computer, so I can guarantee that the definition has no problem. But the proof is very long. I'm afraid that if I also include the proof in this PR, then it may not be so easy to be merged to Mathlib 4. I plan to include the proof of the statement in some other PR.

But that is only my own opinion. I'm not sure if my plan is good. What's your opinion?

Reply:
Please make a separate PR dependent on this one. That one may eventually be split but the reviewers can have a fuller picture

Reply:
> Please make a separate PR dependent on this one. That one may eventually be split but the reviewers can have a fuller picture

I have created a new PR #19404 in which I have proved the key property of Hilbert polynomials. I have also marked #19404 as dependent on #19303.